Add documentation for OPA batch column masking (#22839)
trinodb/trino: docs/src/main/sphinx/security/opa-access-control.md

* - `opa.policy.column-masking-uri`
  - The **optional** URI for fetching column masks - if not set no masking
    is applied. For example, `https://opa.example.com/v1/data/columnMask`.
* - `opa.policy.batch-column-masking-uri`

>   -  The **optional** URI for fetching columns masks in batches - if not set

Reviewer: I'd say something like "if set, it overrides the value of `opa.policy.column-masking-uri` such that the plugin will use batch column masking"

Or something like that?

>   -  The **optional** URI for fetching columns masks in batches; should **not**

Reviewer 2: what you mean by "should" .. can not or must not is better ..

     be used with `opa.policy.column-masking-uri`. For example,

>      `http://opa.example.com/v1/data/batchColumnMasks`.

Reviewer 2: is this the default URI path from an OPA server?

Author: The is no default URI path; the path is determined by the Rego policy, roughly `/v1/data/{package.replace(".", "/")}/{rule}`.
For the example Rego this PR adds the path would be `/v1/data/trino/batchColumnMasking`

Reviewer 2: ok.. still makes this a reasonable example then. Maybe its worth doing /v1/data/example/batchColumnMasking ?

Author: Far point, checked offline with @vagaerg and I think `/v1/data/batchColumnMasks` is not actually a valid path as it has no package (though there may be some OPA config which allows this that I am not aware of).
Updated all paths in the doc.

The same "identity" field may be returned to evaluate column masks under a
different identity.

### Batch column masking

If column masking is enabled, by default, the plugin will fetch each column
mask individually from OPA. When working with very wide tables this
can result in a performance degradation.

> Configuring `opa.policy.batch-column-masks` allows Trino to fetch the masks
> for multiple columns in one requests. The requested column are listed under
> the `action.filterResources` node.

Reviewer: ```suggestion
Configuring `opa.policy.batch-column-masking-uri` allows Trino to fetch the masks
for multiple columns in a single request. The list of requested columns is included in the
request under `action.filterResources`.
```

> An OPA policy supporting batch column masking must return a list containing
> objects with two keys - `viewExpression`, the column mask expression and 
> optional identity, and `index`, the index of the column in the request for
> which the mask applies.

Reviewer: ```suggestion
An OPA policy supporting batch column masking must return a list of objects,
each containing:
- `viewExpression`: an object made of:
    - `expression`: the expression to apply to the column, as a string
    - `identity` (optional): the identity to evaluate the expression as, as a string
- `index`: references the index of the column in the request to which this mask applies
```

Author: I wanted to avoid a list but tbf there are other lists in the same doc so it should be fine

        },
        "softwareStack": {
            "trinoVersion": "434"
        }
    },
    "action": {
        "operation": "GetColumnMask",
        "filterResources": [
            {
                "column": {
                    "catalogName": "sample_catalog",
                    "schemaName": "sample_schema",
                    "tableName": "restricted_table",
                    "columnName": "user_phone",
                    "columnType": "VARCHAR"
                }
            },
            {
                "column": {
                    "catalogName": "sample_catalog",
                    "schemaName": "sample_schema",
                    "tableName": "restricted_table",
                    "columnName": "user_name",
                    "columnType": "VARCHAR"
                }
            }
        ]
    }
}
```

For which the OPA response would be:

```json
[
    {
        "index": 0,
        "viewExpression": {
            "expression": "NULL"
        }
    },
    {
        "index": 1,
        "viewExpression": {
            "expression": "'****' || substring(user_name, -3)",
            "identity": "admin"
        }
    }
]
```

> 

Reviewer: I would add:

```
Making Trino apply expression `NULL` to the column at index 0 (`user_phone`)
and expression `'****' || substring(user_name, -3)`, evaluated as user `admin`
to the column at index 1 (`user_name`)

The same "identity" field may be returned to evaluate column masks under a
different identity.

### Batch column masking

> If column masking is enabled, by default, the plugin will fetch each column

Reviewer 2: ```suggestion
If column masking is enabled, the plugin fetches each column
```

mask individually from OPA. When working with very wide tables this
can result in a performance degradation.

Configuring `opa.policy.batch-column-masking-uri` allows Trino to fetch the masks
for multiple columns in a single request. The list of requested columns is included
in the request under `action.filterResources`.

> If `opa.policy.batch-column-masking-uri` is set it will override the value of

Reviewer 2: ```suggestion
If `opa.policy.batch-column-masking-uri` is set it overrides the value of
```

> `opa.policy.column-masking-uri` such that the plugin will use batch column

Reviewer 2: ```suggestion
`opa.policy.column-masking-uri` so that the plugin uses batch column
```

masking.

An OPA policy supporting batch column masking must return a list of objects,

> each containing:

Reviewer 2: ```suggestion
each containing the following data:

```

> - `viewExpression`: an object made of:

Reviewer 2: maybe this is text is redundant so

```suggestion
- `viewExpression`:
```

    - `expression`: the expression to apply to the column, as a string
    - `identity` (optional): the identity to evaluate the expression as, as a
      string
- `index`: a reference the index of the column in the request to which this mask
  applies

For example, a policy configuring batch column masking may be defined by the

> following script:

Reviewer 2: ```suggestion
following rego script:
```

  column_resource.schemaName == "sample_schema"
  column_resource.tableName == "restricted_table"
  column_resource.columnName == "user_name"
}
```

A batch column masking request is similar to the following example:

```json
{
    "context": {
        "identity": {
            "user": "foo",
            "groups": ["some-group"]
        },
        "softwareStack": {
            "trinoVersion": "434"
        }
    },
    "action": {
        "operation": "GetColumnMask",
        "filterResources": [
            {
                "column": {
                    "catalogName": "sample_catalog",
                    "schemaName": "sample_schema",
                    "tableName": "restricted_table",
                    "columnName": "user_phone",
                    "columnType": "VARCHAR"
                }
            },
            {
                "column": {
                    "catalogName": "sample_catalog",
                    "schemaName": "sample_schema",
                    "tableName": "restricted_table",
                    "columnName": "user_name",
                    "columnType": "VARCHAR"
                }
            }
        ]
    }
}
```

> For which the OPA response would be:

Reviewer 2: The related OPA response is displayed in the following snippet:

```json
[
    {
        "index": 0,
        "viewExpression": {
            "expression": "NULL"
        }
    },
    {
        "index": 1,
        "viewExpression": {
            "expression": "'****' || substring(user_name, -3)",
            "identity": "admin"
        }
    }
]
```

> Making Trino apply expression `NULL` to the column at index 0 (`user_phone`)

Reviewer 2: This is not a sentence.. what are you trying to say?

Author: This was trying to explain the effects of the the returned column masks. Removed.
